src/relayclient/DevRelayClient.ts

```diff
@@ -0,0 +1,254 @@
+    } catch (e) {
+      res = e
+    }
+    const ret = waitFunction(res)
+    if (ret != null) {
+      return ret
+    }
+    await sleep(400)
+  }
+  throw new Error('waitForRelay: timed-out')
+}
+
+interface RunServerReturn {
+  httpServer: HttpServer
+  relayServer: RelayServer
+}
+
+export function runServer (
+  web3provider: provider,
+  relayHub: string,
+  devConfig: DevGSNConfig
+): RunServerReturn {
+  const relayUrl = devConfig.relayUrl ?? `http://localhost:${devConfig.relayListenPort}`
+
+  // TODO: read key-pair from temp file?
+  // (otherwise, we deploy a new relay each time)
+  const keyManager = new KeyManager({ ecdsaKeyPair: KeyManager.newKeypair(), workdir: undefined })
+  const txStoreManager = new TxStoreManager({ inMemory: true })
+
+  // @ts-ignore
+  const relayServer = new RelayServer({
+    web3provider,
+    txStoreManager,
+    keyManager,
+    // owner: relayOwner,
+    hubAddress: relayHub,
+    url: relayUrl,
+    baseRelayFee: devConfig.baseRelayFee,
+    pctRelayFee: devConfig.pctRelayFee,
+    gasPriceFactor: devConfig.gasPriceFactor,
+    devMode: devConfig.devMode
+  })
+  relayServer.on('error', (e: any) => {
+    // console.error('ERR:', e.message)
+  })
+
+  let listenPort = devConfig.relayListenPort
+  if (listenPort === undefined) {
+    const m = relayUrl.match(/(?:(https?):\/\/(\S+?)(?::(\d+))?)?$/)
+    if (m != null) {
+      if (!m[0].startsWith('http')) {
+        throw Error('invalid server URL protocol ' + m[0])
+      }
+      listenPort = (m[3] !== undefined) ? parseInt(m[3]) : (m[0] === 'http' ? 80 : 443)
+    }
+  }
+
+  const httpServer = new HttpServer({
+    port: listenPort,
+    backend: relayServer
+  })
+  httpServer.start()
+  return {
+    httpServer,
+    relayServer
+  }
+}
+
+class DevKnownRelays { // extends  KnownRelaysManager {
+  // noinspection JSUnusedGlobalSymbols
+  getRelaysSorted (): RelayRegisteredEventInfo[] {
+    // @ts-ignore
+    return [this.devRelay]
+  }
+
+  // noinspection JSUnusedGlobalSymbols
+  refresh (): void { // ts-ignore
+  }
+
+  devRelay?: RelayRegisteredEventInfo
+}
+
+export interface DevGSNConfig extends Partial<GSNConfig> {
+  relayOwner: Address
+  relayWorkdir?: string
+  relayListenPort?: number
+  relayUrl?: string // defaults to http://localhost:{relayListenPort}
+
+  // TODO: this is actually relay config
+  baseRelayFee?: number
+  pctRelayFee?: number
+  gasPriceFactor: number
+  devMode?: boolean
+}
+
+export class DevRelayClient extends RelayClient {
+  serverStarted: boolean = false
+  httpServer?: HttpServer
+  relayServer?: RelayServer
+  private readonly devConfig: DevGSNConfig;
+
+  constructor (
+    provider: HttpProvider,
+    devConfig: Partial<DevGSNConfig>,
+    overrideDependencies?: Partial<GSNDependencies>
+  ) {
+    super(provider, devConfig,
+      {
+        ...overrideDependencies,
+        knownRelaysManager: new DevKnownRelays() as KnownRelaysManager
+      })
+    this.devConfig = this.config as DevGSNConfig
```

Comment:
You force-cast `Partial<DevGSNConfig>` to `DevGSNConfig`, that would mean bailing on TypeScript's null-safety. Not so much here, as almost every field in DevGSNConfig is optional anyways. 
This is complicating the understanding of what is going on. May I suggest `DevGSNConfig` to have all fields initialised, like `GSNConfig` does, and create a 'config resolver' method similar to `configureGSN`? 
I think this approach makes it way easier to follow the configuration flow.